### ISSUE TITLE
decimal.Equals, Guid.Equals could be improved

### ISSUE THREAD

**Issue by jamesqo** [dotnet/coreclr#7033]
While looking through some of the 16-byte types today I noticed that some of the `Equals` implementations are quite suboptimal... for example [Guid.Equals](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Guid.cs#L969) has a bunch of `return falses` that the JIT does not merge in the assembly. What's more, it compares ints/shorts at a time when it should ideally be able to compare the whole struct in a couple of instructions.

Same for `Decimal`... in Equals we should be able to just compare the 16 bytes and return true early before doing [all these expensive bit operations](https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/src/palrt/decarith.cpp#L1175).

**Comment by jamesqo** [comment]
Another thing I just found in decimal, though not quite related to this issue: the `Decimal(long)` constructor uses [arithmetic right shift](https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/src/mscorlib/src/System/Decimal.cs#L193) even when we always know that, at that point, `value_copy >= 0`. There should be a cast to ulong beforehand, in case the JIT is not able to detect that and generates more code to treat the sign bit specially.

**Comment by mikedn** [comment]
> There should be a cast to ulong beforehand, in case the JIT is not able to detect that and generates more code to treat the sign bit specially.

Huh?!

**Comment by mikedn** [comment]
> Same for Decimal... in Equals we should be able to just compare the 16 bytes and return true early before doing all these expensive bit operations.

I don't think that will work because the decimal representation isn't guaranteed to be normalized.

**Comment by jamesqo** [comment]
> Huh?!

In [a previous PR](https://github.com/dotnet/coreclr/pull/6434/commits/6c992a6183a8a92453404b0c46b973b9f60625f5#diff-d2a641ec9b77a5ea1cc985bdec3e74efL2358) a couple (many) of extra instructions were being generated for `a % b` if both operands were signed.

```diff
        F7DA                 neg      edx
-       448BC2               mov      r8d, edx
-       41C1F81F             sar      r8d, 31
-       4183E007             and      r8d, 7
-       4403C2               add      r8d, edx
-       4183E0F8             and      r8d, -8
-       412BD0               sub      edx, r8d
-       448BC2               mov      r8d, edx
-       41C1E81F             shr      r8d, 31
-       4403C2               add      r8d, edx
-       418BD0               mov      edx, r8d
-       D1FA                 sar      edx, 1
+       83E207               and      edx, 7
+       D1EA                 shr      edx, 1
```

I am not sure if it would do the same for division/arithmetic left shift, but I know it has to treat the sign bit specially if the number is negative.

> I don't think that will work because the decimal representation isn't guaranteed to be normalized.

Can you explain what you mean? I wasn't saying that they aren't equal if their bits aren't equal (I am aware some bits in Decimal are unused). I was just saying that if their bits are equal then they must be equal.

**Comment by mikedn** [comment]
> In a previous PR a couple (many) of extra instructions were being generated for a % b if both operands were signed.

I don't see how that is related to this, division/modulo is one thing and bit shifting is another thing.

>  I was just saying that if their bits are equal then they must be equal.

But the problem is the other way around. If the bits aren't equal it doesn't mean that the numeric values aren't equal, sometimes they are:
```C#
decimal x = new decimal(32, 0, 0, false, 0);
decimal y = new decimal(320, 0, 0, false, 1);

Console.WriteLine(x == y);

int[] ax = decimal.GetBits(x);
int[] ay = decimal.GetBits(y);

Console.WriteLine(ax.SequenceEqual(ay));
```
The above code prints true/false rather than true/true.

**Comment by jamesqo** [comment]
> If the bits aren't equal it doesn't mean that the numeric values aren't equal

Yep, but if the bits *are* equal it does mean that the numeric values are equal.

**Comment by mikedn** [comment]
> Yep, but if the bits are equal it does mean that the numeric values are equal.

Yes but if the bits aren't equal you still need to use the existing implementation. So what do you gain?
